tests/compile-fail/concurrency/dangling_tls_lib.rs

```diff
@@ -0,0 +1,49 @@
+// ignore-windows: Concurrency on Windows is not supported yet.
+
+//! Check that we catch if a thread local is accessed after the thread has
+//! terminated.
+
+#![feature(thread_local_internals)]
```

Comment:
Why does this test have to use the internals?

Reply:
Ah, probably because otherwise we'd be testing language TLS? Do we have a dangling test for them?

Reply:
Actually, there is no such thing as "pointing to a lib-TLS variable", right? You can just access it with API functions.

We don't remove lib TLS vars on thread exit, but they become inaccessible nevertheless.
What this here really tests is that lib TLS dtors are run, I think? But we already test that with a run-pass test.

I am not sure what this test is covering, but it seems to me what we actually need is a test that takes a ptr to a `#[thread_local]` static inside some thread, exits that thread, and then accesses that ptr. That should be UB. And I think so far I have not seen code in your PR that actually makes this UB, but maybe I missed it.

Reply:
> What this here really tests is that lib TLS dtors are run, I think? But we already test that with a run-pass test.

Actually, yes you are right.

> I am not sure what this test is covering, but it seems to me what we actually need is a test that takes a ptr to a #[thread_local] static inside some thread, exits that thread, and then accesses that ptr. That should be UB. And I think so far I have not seen code in your PR that actually makes this UB, but maybe I missed it.

I have tried to add code that deallocates thread local allocations. However, if I understand correctly, there is no way to delete an allocation from [`tcx.alloc_map`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/interpret/struct.AllocMap.html). So, it seems that the best I can do before https://github.com/rust-lang/rust/issues/70685 gets fixed is to at least mark the memory as unitialized (and open an issue to track this). Or, do you have better suggestions?

With marking the memory as uninitialized, the assignment to `x` goes through, but assignment to `y` fails with UB:
```rust
#![feature(thread_local)]

use std::thread;

#[thread_local]
static mut A: u8 = 0;

struct Sender(*const u8);

unsafe impl Send for Sender {}

fn main() {
    let handle = thread::spawn(|| {
        let ptr = unsafe { &mut A as *mut u8 };
        assert_eq!(unsafe { *ptr }, 0);
        unsafe { *ptr = 5; }
        Sender(ptr)
    });
    let ptr = handle.join().unwrap().0;
    let x = unsafe { *ptr };
    let y = x + 1;
}
```

Reply:
Let's postpone this until after https://github.com/rust-lang/rust/issues/70685 lands -- but let's have an issue tracking it.